### PR TITLE
En 4444/integrate tx coordinator into metachain

### PR DESCRIPTION
**Pull request by sasurobert**
Transaction coordinator for metachain. Process, commit, create, restore and remove miniblocks. Process transaction where destination is the metachain.
Add verification to txValidator to not let bad transactions to metachain to enter the system.
Verify genesis that there are no addresses ending in "ff".
Limit the creation of smart contract calls from shard to metablock to only one full miniblock - so upon reaching gasLimit, no more transactions to be processed.

### CODE REVIEW

File: process/block/argProcessor.go [hunk 1]
```diff
@@ -42,5 +42,6 @@ type ArgShardProcessor struct {
 // new instances of meta processor
 type ArgMetaProcessor struct {
 	ArgBaseProcessor
-	DataPool dataRetriever.MetaPoolsHolder
+	DataPool      dataRetriever.MetaPoolsHolder
+	TxCoordinator process.TransactionCoordinator
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/argProcessor.go]: done

File: core/address.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		make([]byte, hooks.NumInitCharactersForScAddress-hooks.VMTypeLen))
+	if isSCAddress {
+		return true
+	}
+
+	return false
+}
+
+// IsMetaChainShardId verifies if the identifier is of type metachain
+func IsMetaChainShardId(identifier []byte) bool {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/address.go]: done.

File: process/block/metablock.go [hunk 3]
```diff
@@ -135,14 +149,21 @@ func (mp *metaProcessor) ProcessBlock(
 		mp.headersCounter.getNumShardMBHeadersTotalProcessed(),
 	)
 
+	mp.txCoordinator.CreateBlockStarted()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 4]
```diff
@@ -195,14 +221,129 @@ func (mp *metaProcessor) ProcessBlock(
...
+			continue
+		}
+		shardHeader, ok := hdrInfo.hdr.(*block.Header)
+		if !ok {
+			continue
+		}
+
+		lastHdr, err := mp.getLastNotarizedHdr(shardInfo.ShardID)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 5]
```diff
@@ -195,14 +221,129 @@ func (mp *metaProcessor) ProcessBlock(
...
+		if mbHdr.SenderShardID != miniBlock.SenderShardID {
+			return process.ErrHeaderBodyMismatch
+		}
+	}
+
+	return nil
+}
+
+func (mp *metaProcessor) verifyCrossShardMiniBlockDstMe(header *block.MetaBlock) error {
+	miniBlockMetaHashes, err := mp.getAllMiniBlockDstMeFromMeta(header)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: I choose dstMeta, as smart contract results generated by meta will be in this verification as well.

File: integrationTests/testSyncNode.go [hunk 6]
```diff
@@ -101,6 +101,7 @@ func (tpn *TestProcessorNode) initBlockProcessorWithSync() {
 		arguments := block.ArgMetaProcessor{
 			ArgBaseProcessor: argumentsBase,
 			DataPool:         tpn.MetaDataPool,
+			TxCoordinator:    &mock.TransactionCoordinatorMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testSyncNode.go]: done

File: cmd/node/factory/structs.go [hunk 7]
```diff
@@ -1851,6 +1852,8 @@ func newMetaBlockProcessor(
 	arguments := block.ArgMetaProcessor{
 		ArgBaseProcessor: argumentsBaseProcessor,
 		DataPool:         data.MetaDatapool,
+		//TODO add real transaction coordinator here
+		TxCoordinator: &mock.TransactionCoordinatorMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done

File: integrationTests/multiShard/smartContract/testInitializer.go [hunk 8]
```diff
@@ -853,7 +853,8 @@ func createMetaNetNode(
 			RequestHandler:  requestHandler,
 			Core:            &mock.ServiceContainerMock{},
 		},
-		DataPool: dPool,
+		DataPool:      dPool,
+		TxCoordinator: &mock.TransactionCoordinatorMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/smartContract/testInitializer.go]: done

File: integrationTests/testProcessorNode.go [hunk 9]
```diff
@@ -509,6 +509,7 @@ func (tpn *TestProcessorNode) initBlockProcessor() {
 		arguments := block.ArgMetaProcessor{
 			ArgBaseProcessor: argumentsBase,
 			DataPool:         tpn.MetaDataPool,
+			TxCoordinator:    &mock.TransactionCoordinatorMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testProcessorNode.go]: done

File: process/block/metablock_test.go [hunk 10]
```diff
@@ -38,7 +38,8 @@ func createMockMetaArguments() blproc.ArgMetaProcessor {
 			RequestHandler:        &mock.RequestHandlerMock{},
 			Core:                  &mock.ServiceContainerMock{},
 		},
-		DataPool: mdp,
+		DataPool:      mdp,
+		TxCoordinator: &mock.TransactionCoordinatorMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock_test.go]: done

File: process/block/metablock.go [hunk 11]
```diff
@@ -21,9 +21,9 @@ import (
 // metaProcessor implements metaProcessor interface and actually it tries to execute block
 type metaProcessor struct {
 	*baseProcessor
-	core               serviceContainer.Core
-	dataPool           dataRetriever.MetaPoolsHolder
-	//TODO: add	txCoordinator process.TransactionCoordinator
+	core          serviceContainer.Core
+	dataPool      dataRetriever.MetaPoolsHolder
+	txCoordinator process.TransactionCoordinator
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 12]
```diff
@@ -44,6 +44,9 @@ func NewMetaProcessor(arguments ArgMetaProcessor) (*metaProcessor, error) {
 	if arguments.DataPool.ShardHeaders() == nil || arguments.DataPool.ShardHeaders().IsInterfaceNil() {
 		return nil, process.ErrNilHeadersDataPool
 	}
+	if arguments.TxCoordinator == nil || arguments.TxCoordinator.IsInterfaceNil() {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 13]
```diff
@@ -76,6 +79,7 @@ func NewMetaProcessor(arguments ArgMetaProcessor) (*metaProcessor, error) {
 		baseProcessor:  base,
 		dataPool:       arguments.DataPool,
 		headersCounter: NewHeaderCounter(),
+		txCoordinator:  arguments.TxCoordinator,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 14]
```diff
@@ -195,14 +221,129 @@ func (mp *metaProcessor) ProcessBlock(
...
+	for hash := range crossMiniBlockHashes {
+		if _, ok := miniBlockMetaHashes[hash]; !ok {
+			return process.ErrCrossShardMBWithoutConfirmationFromMeta
+		}
+	}
+
 	return nil
 }
 
+func (mp *metaProcessor) getAllMiniBlockDstMeFromMeta(metaHdr *block.MetaBlock) (map[string][]byte, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply process/block/metablock.go]: I see you actually went with @SebastianMarian suggestion. Was this the intended change? Just making sure this is what you wanted

[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: @SebastianMarian suggestion was better, as I introduced miniblockheaders for better clarity.

File: process/block/metablock.go [hunk 15]
```diff
@@ -945,29 +1240,7 @@ func (mp *metaProcessor) createShardInfo(
...
-	log.Debug(fmt.Sprintf("time elapsed to ordered %d hdrs: %v sec\n", len(orderedHdrs), timeAfter.Sub(timeBefore).Seconds()))
-
-	if err != nil {
-		return nil, err
-	}
-
-	log.Info(fmt.Sprintf("creating shard info has been started: have %d hdrs in pool\n", len(orderedHdrs)))
+	log.Info(fmt.Sprintf("creating shard info has been started \n"))
 
 	// save last committed header for verification
```
Comment (SebastianMarian) [inline process/block/metablock.go]: You can delete lines between 1245-1254 and line 1237

File: process/block/metablock.go [hunk 16]
```diff
@@ -945,29 +1240,7 @@ func (mp *metaProcessor) createShardInfo(
 		return nil, process.ErrAccountStateDirty
 	}
 
-	if !haveTime() {
```
Comment (SebastianMarian) [inline process/block/metablock.go]: Just think if these prints related to time is up / time elapsed should be moved in createAndProcessCrossMiniBlocksDstMe

File: process/block/metablock.go [hunk 17]
```diff
@@ -933,7 +1229,6 @@ func (mp *metaProcessor) checkAndProcessShardMiniBlockHeader(
 }
 
 func (mp *metaProcessor) createShardInfo(
-	maxItemsInBlock uint32,
 	round uint64,
 	haveTime func() bool,
```
Comment (SebastianMarian) [inline process/block/metablock.go]: haveTime is not used in this method, do we want this?

File: process/block/metablock.go [hunk 18]
```diff
@@ -981,94 +1254,42 @@ func (mp *metaProcessor) createShardInfo(
...
-			mp.hdrsForCurrBlock.mutHdrsForBlock.Unlock()
-			return shardInfo, nil
 		}
 
-		if len(shardData.ShardMiniBlockHeaders) == len(orderedHdrs[index].MiniBlockHeaders) {
-			shardInfo = append(shardInfo, shardData)
-			mp.hdrsForCurrBlock.hdrHashAndInfo[string(orderedHdrHashes[index])] = &hdrInfo{hdr: orderedHdrs[index], usedInBlock: true}
-		}
+		shardInfo = append(shardInfo, shardData)
+		mp.hdrsForCurrBlock.hdrHashAndInfo[string(hdrHash)].usedInBlock = true
```
Comment (SebastianMarian) [inline process/block/metablock.go]: This is already set to true. You can remove this line

File: process/block/metablock.go [hunk 19]
```diff
@@ -1095,9 +1316,13 @@ func (mp *metaProcessor) CreateBlockHeader(bodyHandler data.BodyHandler, round u
 
 	defer func() {
 		go mp.checkAndRequestIfShardHeadersMissing(round)
+
+		mp.blockSizeThrottler.Add(
```
Comment (SebastianMarian) [inline process/block/metablock.go]: This should be called only when block has been created with success

File: data/block/metaBlock.go [hunk 20]
```diff
@@ -421,30 +435,24 @@ func (m *MetaBlock) SetTxCount(txCount uint32) {
...
-// IntegrityAndValidity return true as block is nil for metablock.
-func (m *MetaBlockBody) IntegrityAndValidity() error {
-	return nil
-}
-
-// IsInterfaceNil returns true if there is no value under the interface
-func (m *MetaBlockBody) IsInterfaceNil() bool {
-	if m == nil {
-		return true
+	for _, val := range m.MiniBlockHeaders {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock.go]: not right now, but maybe in near future. There are some usecases.

File: data/block/metaBlock.go [hunk 21]
```diff
@@ -43,38 +43,35 @@ type PeerData struct {
...
 type ShardData struct {
-	ShardId               uint32                 `capid:"0"`
+	ShardID               uint32                 `capid:"0"`
 	HeaderHash            []byte                 `capid:"1"`
 	ShardMiniBlockHeaders []ShardMiniBlockHeader `capid:"2"`
 	TxCount               uint32                 `capid:"3"`
 }
 
 // MetaBlock holds the data that will be saved to the metachain each round
 type MetaBlock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock.go]: Any suggestion how to resolve ?

File: process/block/preprocess/transactions.go [hunk 22]
```diff
@@ -441,6 +423,19 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
+		txs.shardCoordinator.SelfId(),
+		sharding.MetachainShardId,
+		txSpaceRemained,
+		haveTime,
+		round)
+
+	if err == nil && len(miniBlock.TxHashes) > 0 {
+		txSpaceRemained -= len(miniBlock.TxHashes)
+		miniBlocks = append(miniBlocks, miniBlock)
+		newMBAdded = true
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/transactions.go]: done

File: core/address.go [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+// IsSmartContractOnMetachain verifies if an address is smart contract on metachain
+func IsSmartContractOnMetachain(identifier []byte, rcvAddress []byte) bool {
+	if !IsMetachainIdentifier(identifier) {
+		return false
+	}
+
+	if !IsSmartContractAddress(rcvAddress) {
+		return false
+	}
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/address.go]: done

File: process/block/metablock.go [hunk 24]
```diff
@@ -195,14 +216,92 @@ func (mp *metaProcessor) ProcessBlock(
...
 	return nil
 }
 
+func (mp *metaProcessor) verifyCrossShardMiniBlockDstMe(header *block.MetaBlock) error {
+	miniBlockShardsHashes, err := mp.getAllMiniBlockDstMeFromShards(header)
+	if err != nil {
+		return err
+	}
+
+	crossMiniBlockHashes := header.GetMiniBlockHeadersWithDst(mp.shardCoordinator.SelfId())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 25]
```diff
@@ -362,7 +474,175 @@ func (mp *metaProcessor) CreateBlockBody(round uint64, haveTime func() bool) (da
...
+	mbFromMe := mp.txCoordinator.CreateMbsAndProcessTransactionsFromMe(
+		uint32(maxTxSpaceRemained),
+		uint32(maxMbSpaceRemained),
+		round,
+		haveTime)
+
+	if len(mbFromMe) > 0 {
+		miniBlocks = append(miniBlocks, mbFromMe...)
+	}
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 26]
```diff
@@ -362,7 +474,175 @@ func (mp *metaProcessor) CreateBlockBody(round uint64, haveTime func() bool) (da
...
+	miniBlocks := make(block.MiniBlockSlice, 0)
+	txsAdded := uint32(0)
+	hdrsAdded := uint32(0)
+	lastPushedHdr := make(map[uint32]data.HeaderHandler, mp.shardCoordinator.NumberOfShards())
+
+	orderedHdrs, orderedHdrHashes, sortedHdrPerShard, err := mp.getOrderedHdrs(round)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 27]
```diff
@@ -362,7 +474,175 @@ func (mp *metaProcessor) CreateBlockBody(round uint64, haveTime func() bool) (da
...
+			break
+		}
+
+		hdr := orderedHdrs[i]
+		lastHdr, ok := lastPushedHdr[hdr.ShardId].(*block.Header)
+		if !ok {
+			continue
+		}
+
+		isFinal, _ := mp.isShardHeaderValidFinal(orderedHdrs[i], lastHdr, sortedHdrPerShard[hdr.ShardId])
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go (hunk 15 above)
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go (hunk 17 above)
Comment (sasurobert) [reply process/block/metablock.go]: yes. as we have haveTime in createBlockBody - header creation should not depend on time.

File: process/block/metablock.go (hunk 18 above)
Comment (sasurobert) [reply process/block/metablock.go]: I wanted to have this set to true, only when body and header is processed.

File: process/block/metablock.go (hunk 19 above)
Comment (sasurobert) [reply process/block/metablock.go]: done

File: data/block/metaBlock.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply data/block/metaBlock.go]: Just remove capn behind line from import "github.com/glycerine/go-capnproto"

File: core/address.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply core/address.go]: Yous should do this also in IsSmartContractAddress

File: process/block/metablock.go (hunk 17 above)
Comment (SebastianMarian) [reply process/block/metablock.go]: Than you should remove this parameter as it is not used anywhere inside method

File: process/block/metablock.go (hunk 19 above)
Comment (SebastianMarian) [reply process/block/metablock.go]: You should add this also in: if bodyHandler == nil || bodyHandler.IsInterfaceNil() {...

File: process/block/metablock.go (hunk 18 above)
Comment (SebastianMarian) [reply process/block/metablock.go]: It is created by default with true in createAndProcessCrossMiniBlocksDstMe method

File: data/block/metaBlock.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock.go]: done

File: core/address.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply core/address.go]: done

File: process/block/metablock.go (hunk 17 above)
Comment (sasurobert) [reply process/block/metablock.go]: will be removed in another pull request, it is already open.

File: process/block/metablock.go (hunk 18 above)
Comment (sasurobert) [reply process/block/metablock.go]: done.

Comment (sasurobert) [reply process/block/metablock.go]: removed

File: process/block/metablock.go (hunk 19 above)
Comment (sasurobert) [reply process/block/metablock.go]: throttler is called only if there was a body, and if it was correctly processed.

Comment (sasurobert) [reply process/block/metablock.go]: and that check is added in createHeader

File: core/address.go [hunk 28]
```diff
@@ -37,6 +41,10 @@ func IsMetachainIdentifier(identifier []byte) bool {
 
 // IsSmartContractOnMetachain verifies if an address is smart contract on metachain
 func IsSmartContractOnMetachain(identifier []byte, rcvAddress []byte) bool {
+	if len(rcvAddress) <= hooks.NumInitCharactersForScAddress+numInitCharactersForOnMetachainSC {
```
Comment (ccorcoveanu) [inline core/address.go]: `numInitCharactersForOnMetachainSC` - I can't really decode what it means: `ForOnMetachain`, is that correct?

File: core/address.go [hunk 29]
```diff
@@ -10,6 +10,10 @@ const numInitCharactersForOnMetachainSC = 5
 
 // IsSmartContractAddress verifies if a set address is of type smart contract
 func IsSmartContractAddress(rcvAddress []byte) bool {
+	if len(rcvAddress) <= hooks.NumInitCharactersForScAddress {
```
Comment (ccorcoveanu) [inline core/address.go]: Can this function be called with something else than a valid address (32 bytes)?

**Pull request merged**